book/src/e2e-testing.md

```diff
@@ -0,0 +1,81 @@
+# End to End Testing for CSI Storage Plugins
+
+Currently, csi-sanity exists to help test compliance with the spec. But e2e testing of plugins is needed as well to provide plugin authors and users stronger guarantees that their plugin "just works" in kubernetes.
+
+## Setting up End to End tests for your CSI Plugin
+
+### Prerequisites:
+ * A Kubernetes v1.12+ Cluster
+ * [Kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl/#install-kubectl)
+
+There are two ways to run end-to-end tests for your CSI Plugin
+ 1) use [Kubernetes E2E Tests](https://github.com/kubernetes/kubernetes/tree/master/test/e2e/storage/external), by providing a DriverDefinition YAML file via a parameter.
+ 2) import the in-tree storage tests and run them using `go test`. 
+ 
+ This doc will cover how to run the E2E tests using the second method. 
+ 
+### Note: 
+ 
+ In some cases you would not be able to use the first method, in running e2e tests by just providing a YAML file defining your CSI plugin.
+
+ * For example the [NFS CSI plugin](https://github.com/kubernetes-csi/csi-driver-nfs) does not support dynamic provisoning, so we would want to skip those and run only pre-provisioned tests. For such cases, you would need to write your own testdriver, which is discussed below. 
+
+In-tree storage e2e tests could be used to test CSI storage plugins. Your repo could be setup similar to how this [NFS CSI plugin](https://github.com/mathu97/csi-driver-nfs/) is setup, where the testfiles are in a test directory and the main test file is in the cmd directory. 
+
+To be able import the in-tree storage tests, CSI plugin authors would need to implement a [testdriver](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/testdriver.go#L31) for their CSI plugin. The testdriver provides required functionality that would help setup testcases for a particular plugin. 
```

Comment:
It should be noted somewhere that the version of the test library that you import is tied to the Kubernetes version.  For example, the 1.13+ version of the e2e tests runs the raw block tests by default, since the feature went beta in 1.13.  If you ran it against a 1.12 cluster, it would fail if you didn't enable the alpha feature.